addons/test_mail/tests/test_performance.py

```diff
@@ -696,6 +696,119 @@ def test_complex_tracking_template(self):
         self.assertEqual(rec1.message_ids[2].notified_partner_ids, self.partners | self.user_portal.partner_id)
         self.assertEqual(len(rec1.message_ids), 3)
 
+    @mute_logger('odoo.tests', 'odoo.addons.mail.models.mail_mail', 'odoo.models.unlink')
+    @users('emp')
+    @warmup
+    def test_message_format(self):
+        """Test performance of `_message_read_dict_postprocess` and of
+        `message_format` with multiple messages with multiple attachments,
+        different authors, various notifications, and different tracking values.
+        Those messages might not make sense functionally but they are crafted to
+        cover as much of the code as possible in regard to number of queries.
+        """
+        name_field = self.env['ir.model.fields']._get(self.umbrella._name, 'name')
+        customer_id_field = self.env['ir.model.fields']._get(self.umbrella._name, 'customer_id')
+
+        messages = self.env['mail.message'].sudo().create([{
```

Comment:
Isn't it possible to shorten a bit data creation ? 80 lines for 2 messages, looks like JS tests. Some data is redundant, could be done using lists, ...

Reply:
I don't really see how to reduce it while keeping the current coverage, but if you have the time to give me hints I'm open to shorten it.

Reply:
Example: 
```
            'notification_ids': [
                (0, 0, {
                    'res_partner_id': self.partners[3].id,
                    'notification_type': 'inbox',
                }),
                (0, 0, {
                    'res_partner_id': self.partners[4].id,
                    'notification_type': 'email',
                    'notification_status': 'exception',
                }),
                (0, 0, {
                    'res_partner_id': self.partners[6].id,
                    'notification_type': 'email',
                    'notification_status': 'exception',
                }),
            ],
```

```
'notification_ids': [ (0, 0, {
                    'res_partner_id': self.partners[partner_index].id,
                    'notification_type': n_type,
                    'notification_status': n_status,
                }) for (partner_index, n_type, n_status) in [(3, 'inbox', 'ready'), (4, 'email', 'exception'),  (5, 'email, exception)', ]]
```
(Not perfect, but an example of a way to shorten it)